File: Queue/python-implementation/queue.py
class Person(object):
	def __init__(self, name, age):
		self.name = name
		self.age = age
		self.next = None


class queuePerson(object):
	def __init__(self):
		self.head = None
		self.temp = None

	def enQueue(self, name, age):
		per = Person(name, age)
		if self.head is None:
			self.head = per
			self.temp = per
			return self.head
		self.temp.next = per
		self.temp = per
		return self.head
	
	def front(self):	
		return self.head.name, self.head.age

	def deQueue(self):
		self.head = self.head.next
	
	def isEmpty(self):
		if self.head is None:
			return "Empty"
		return "Not Empty"
		
def main():
	pers = queuePerson()
	head = pers.enQueue("Tenkorang", 22) 
	head = pers.enQueue("Joseph", 23) 
	head = pers.enQueue("Ofori", 21) 
	head = pers.enQueue("Adelaide", 20) 

	head = pers.front()
	print(head)

	pers.deQueue()

	head = pers.front()
	print(head)


	head = pers.isEmpty()
	print(head)		

if __name__ == '__main__':
	main()		

					
	
						
		
		

File: Queue/C-implementation/queue.c
/*
In a queue data structure, insertion is done at the bottom while deletion is done from the 
front of the structure. this is so because in a queue whatever goes in first comes out first like 
it os experienced in real life
*/

#include <stdio.h>
#include <stdlib.h>

struct Employee{
	int age;
	double salary;
	struct Employee* next;
};

struct Employee* head = NULL;
struct Employee* temp = NULL;

struct Employee* createEmployee(int age, double salary){
	struct Employee* newEmployee = (struct Employee*)malloc(sizeof(struct Employee));

	newEmployee->age = age;
	newEmployee->salary = salary;
	newEmployee->next = NULL;
	return newEmployee;
}

// The temp variable stores the address of the recently added node to make insertion of other nodes 
// easier ans faster 
struct Employee* enQueue(int age, double salary){
	struct Employee* emp = createEmployee(age,salary);
	if (head == NULL){
		head = emp;
		temp = emp;
		return head;
	}
	temp->next = emp;
	temp = emp;
	return head;	
}

struct Employee* deQueue(struct Employee* head){
	struct Employee* temp = head;
	if (head == NULL){
		return 0;
	}
	head = temp->next;
	free(temp);
	return head;
}

void Print(struct Employee* head){
	if (head == NULL){
		return;
	}
	printf("%d  %.1lf\n",head->age, head->salary);
	Print(head->next);
}

void front(struct Employee* head){
	printf("%d  %.1lf\n", head->age, head->salary);
}
int isEmpty(struct Employee* head){
	int check = 0;
	if (head == NULL){
		check = 1;
	}
	return check;
}

int main(){
	struct Employee* head;
	head = enQueue(22, 20000);
	head = enQueue(23, 30000);
	head = enQueue(24, 40000);
	head = enQueue(25, 50000);

	head = deQueue(head);

	front(head);
	printf("%d\n",isEmpty(head));
	//Print(head);
}